File: R/spatial_join.R
#' @title Spatial Join
#' @description Spatially joins two \code{tbl}.
#' @param x \code{tbl} to join.
#' @param y \code{tbl} to join.
#' @param latitude Latitude variable.
#' @param longitude Longitude variable.
#' @param polygon Polygon variable as usually provided by
#' \code{\link{spark_read_geojson}}.
#' @return \code{tbl} resulting from the spatial join.
#' @note The current implementation performs a cartesian product with a
#' subsequent filter since generic joins are not currently available in
#' \pkg{dplyr} (see https://github.com/tidyverse/dplyr/issues/2240).
#' Spark SQL's Catalyst optimizer seems to be able to optimize this
#' in a way it is not necessary to compute a full cartesian product.
#' @references
#' \url{https://github.com/harsha2010/magellan}
#' @import dplyr
#' @export
spatial_join <- function(x, y, latitude, longitude, polygon) {
  latitude <- enquo(latitude)
  longitude <- enquo(longitude)
  polygon <- enquo(polygon)

  # inner join with subsequent filter since generic joins are not available
  # yet in dplyr: https://github.com/tidyverse/dplyr/issues/2240

  x %>%
    mutate(dummy = TRUE) %>% # https://github.com/tidyverse/dplyr/issues/1841
    inner_join(y %>% mutate(dummy = TRUE)) %>%
    mutate(is_within = within(point(!!latitude, !!longitude), !!polygon)) %>%
    filter(is_within) %>%
    select(-one_of("dummy", "is_within"))
}

File: java/SpatialJoin.scala
package com.miraisolutions.spark.geo

import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.functions._
import org.apache.spark.sql.magellan.dsl.expressions._

object SpatialJoin {
  private val pointColumnName = "point"
  private val indexColumn = "index"

  def join(left: DataFrame, right: DataFrame, latitude: String, longitude: String, polygon: String,
           dropPolygon: Boolean = true): DataFrame = {
    val df = left
      .withColumn(pointColumnName, point(col(longitude), col(latitude)))
      .join(right)
      .where(col(pointColumnName) within col("polygon"))
      .drop(pointColumnName)

    if(dropPolygon) {
      df.drop(polygon, indexColumn)
    } else {
      df
    }
  }
}
